core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveEmptyExceptBranches.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ListMultimap;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.AggregationNode.Step;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.ExceptNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.List;
+import java.util.Optional;
+
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isEmpty;
+import static io.trino.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.trino.sql.planner.plan.Patterns.except;
+
+/**
+ * Removes branches from an ExceptNode that are guaranteed to produce 0 rows.
+ *
+ * <code>A EXCEPT E1 EXCEPT E2 ... EXCEPT En</code> is equivalent to <code>A - (E1 + ... + En)</code>.
```

Comment:
There is a potential "race condition" between this rule and the `EvaluateEmptyExcept` rule in the case when `A` is empty and all of `Ei` are empty. I think we need to resolve it.

I generally always try to avoid situations like this, even if the overall result is the same for both paths (just to make it easier to analyze the flow in the future). In this case however, there is no guarantee that results will be the same. That might change in the future when we add full "empty source folding". But currently, the rule `EvaluateEmptyExcept` gives a better plan.

I propose that we add a condition that `A` is not empty in this rule (with a comment), or merge both rules.

Sorry for noting things late again.

Reply:
You're right -- I misremembered. I thought we had a rule that would convert an Agg+Group By over empty values to empty value.